wgsl/index.bs

```diff
@@ -931,9 +941,230 @@ a result value, such as the non-result-value effects of its subexpressions.
+
+Ideal types support a broad range of operations including arithmetic and builtin functions, but
+subject to <dfn noexport>pragmatic limits</dfn>.
+* the mathematically correct result for an [=IdealInteger=] operation must be within [-2<sup>63</sup>, 2<sup>63</sup>-1].
+* the mathematically correct result for an [=IdealReal=] operation must be within the finite range of
+    a [[!IEEE-754|IEEE-754]] binary64 (double precision) number, and may be approximated as if computed by
+    double-precision arithemtic.
+
+A [=shader-creation error=] results if the operations on ideal values violate these pragmatic ranges.
+
+<table class='data'>
+  <caption>Expressions on ideal types</caption>
+  <thead>
+    <tr><th>Precondition<th>Conclusion<th>Description
+  </thead>
+  <tr algorithm="ideal negation">
+    <td>|e|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>`-`|e|`:` |T|
+        <td>Negation: Evaluates to 0 - |e|.
+            Result must satisfy [=pragmatic limits=].
+
+  <tr algorithm="ideal parenthesization">
+    <td>|e|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>`(`|e|`)` : |T|
+    <td>Parenthesization: Evaluates to |e|.
+
+  <tr algorithm="ideal add">
+    <td>|e1|, |e2|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>|e1| `+` |e2| : |T|
+    <td>Parenthesization: Evaluates to |e1| + |e2|, subject to [=pragmatic limits=].
+
+  <tr algorithm="ideal minus">
+    <td>|e1|, |e2|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>|e1| `-` |e2| : |T|
+    <td>Parenthesization: Evaluates to |e1| - |e2|, subject to [=pragmatic limits=].
+
+  <tr algorithm="ideal multiply">
+    <td>|e1|, |e2|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>|e1| `*` |e2| : |T|
+    <td>Parenthesization: Evaluates to |e1| &times; |e2|, subject to [=pragmatic limits=].
+
+  <tr algorithm="ideal integer divide">
+    <td>|e1|, |e2|: [=IdealInteger=]<br>
+    <td>|e1| `/` |e2| : [=IdealInteger=]
+    <td>Parenthesization: Evaluates to integer division |e1| / |e2|, subject to [=pragmatic limits=].<br>
+        Example:  In integer division: 5 / 2 = 2
+
+  <tr algorithm="ideal real divide">
+    <td>|e1|, |e2|: [=IdealReal=]<br>
+    <td>|e1| `/` |e2| : [=IdealReal=]
+    <td>Parenthesization: Evaluates to real division |e1| / |e2|, subject to [=pragmatic limits=].<br>
+        Example:  In real division: 5 / 2 = 2.5
+</table>
+
+TODO:
+* % operation
+* comparison
+* bitwise | & ^ only if value is non-negative
+    * Rationale?: negative values are problematic because you get infinite numbers of 1 bits.
+* bitwise shift only if e1 and e2 are non-negative, and no modulo clamping for e2
+* builtin functions:
+    * select with numeric arguments
+    * Floating builtins: all of them (scalar only)
+        * frexp, modf need adaptation to return IdealReal components
+    * Integer builtins: all fo them (scalar only), except:
+        * countOneBits only works when non-negative
+        * reverseBits is not generalizable
+
+
```

Comment:
Suggest adding a note that this very likely entails ideal expression being evaluated in the WGSL compiler.